pytorch_toolkit/formula_recognition/im2latex/models/text_recognition_heads/attention_based.py

```diff
@@ -0,0 +1,280 @@
+
+            logits.append(logit)
+        logits = torch.stack(logits, dim=1)  # [B, MAX_LEN, out_size]
+        targets = torch.max(torch.log(logits).data, dim=2)[1]
+        return logits, targets
+
+    def decode_with_bs(self, h, c, O_t, row_enc_out, b_size, device):
+        max_len = self.max_len
+        # initializing empty list of candidates
+        # at first we have only one candidate (START_TOKEN) per image with score = 1
+        candidates_list = []
+        candidates_list.append(Candidate(
+            score=0.,
+            dec_state=(h, c),
+            Out_t=O_t,
+            targets=[START_TOKEN],
+            logits=[]
+        ))
+        # the first prediction is empty with score = 1
+        logits = [[]] * b_size
+        targets = [[]] * b_size
+
+        for _ in range(max_len):
+            # pushing every candidate into step decoding to get probable tracks
+            all_candidates = self.decode_with_bs_step(
+                row_enc_out, candidates_list, device)
+            all_candidates.sort(key=lambda cand: cand.score)
+            candidates_list = all_candidates[:self.beam_width]
+
+        logits = candidates_list[0].logits
+        targets = candidates_list[0].targets[1:]
+        logits = torch.stack(logits, dim=0)
+        logits = logits.unsqueeze(dim=0)
+        logits = logits.to(device)
+        targets = torch.tensor(targets).to(device)
+        targets = targets.unsqueeze(dim=0)
+
+        return logits, targets
+
+    def encode(self, encoded_imgs):
+        encoded_imgs = encoded_imgs.permute(0, 2, 3, 1)  # [B, H', W', LSTM_INP_CHANNELS]
+        # Prepare data for Row Encoder
+        # poccess data like a new big batch
+        B, H, W, out_channels = encoded_imgs.size()
+
+        encoded_imgs = encoded_imgs.contiguous().view(B*H, W, out_channels)
+
+        # prepare init hidden for each row
+        init_hidden_h = self.V_h_0.unsqueeze(
+            1).expand(self.V_h_0.shape[0], B*H, self.V_h_0.shape[1]).contiguous()
+        init_hidden_c = self.V_c_0.unsqueeze(
+            1).expand(self.V_h_0.shape[0], B*H, self.V_h_0.shape[1]).contiguous()
+        init_hidden = (init_hidden_h, init_hidden_c)
+
+        # Row Encoder
+        self.rnn_encoder.flatten_parameters()
+        row_enc_out, (h, c) = self.rnn_encoder(encoded_imgs, init_hidden)
+        # row_enc_out [B*H, W, enc_rnn_h]
+        # hidden: [2, B*H, enc_rnn_h]
+        row_enc_out = row_enc_out.view(B, H, W, self.dec_rnn_h)  # [B, H, W, dec_rnn_h]
+        h = h.view(2, B, H, self.enc_rnn_h)
+        c = c.view(2, B, H, self.enc_rnn_h)
+        return row_enc_out, h, c
+
+    def step_decoding(self, h, c, output, enc_out, tgt):
+        """Runing one step decoding"""
+
+        prev_y = self.embedding(tgt).squeeze(1)  # [B, emb_size]
+        inp = torch.cat([prev_y, output], dim=1)  # [B, emb_size+enc_rnn_h]
+        h_t, c_t = self.rnn_decoder(inp, (h, c))
+
+        context_t, attn_scores = self._get_attn(enc_out, h)
```

Comment:
why attention is used after rnn_decoder step???

Reply:
In other words, why attention is not used by decoder?

Reply:
As I understand, attention is used only in calculating context vector, for details see [decoder section of the section 3](https://arxiv.org/pdf/1609.04938.pdf)

Reply:
and context vector goes to decoder as a part of its input. Is not it?

Reply:
@LeonidBeynenson please have a look